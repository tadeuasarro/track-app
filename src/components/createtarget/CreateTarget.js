import PropTypes from 'prop-types';
import './createtarget.css';

const CreateTarget = ({ target }) => {
  const handleCreateClick = () => {
    console.log('Create a target');
  };

  const handleDeleteClick = () => {
    console.log('Deleting a target');
  };

  if (target) {
    return (
      <button type="button" onClick={handleDeleteClick()} className="delete-target">Delete target</button>
    );
  }

  return (
    <div>
      <input className="target-input" placeholder="Input target here" />
      <button onClick={handleCreateClick()} type="button" className="create-target">Set a new target</button>
    </div>
  );
};

CreateTarget.propTypes = {
  target: PropTypes.number.isRequired,
};

export default CreateTarget;
