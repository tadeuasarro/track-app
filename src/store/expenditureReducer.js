import { SET_EXPENDITURES } from '../actions/expenditure';
import filterExpenditures from '../helpers/filterExpenditures';

const initialState = {
  expenditures: [],
  summary: false,
};

const expenditureReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_EXPENDITURES:
      return {
        expenditures: action.payload,
        summary: filterExpenditures(action.payload),
      };
    default:
      return state;
  }
};

export default expenditureReducer;
