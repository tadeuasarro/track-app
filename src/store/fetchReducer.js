import { SET_LOADING, SET_ERRORS } from '../actions/fetch';

const initialState = {
  loading: false,
  errors: false,
};

const fetchReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_LOADING:
      return {
        loading: true,
        errors: false,
      };
    case SET_ERRORS:
      return {
        loading: false,
        errors: action.payload,
      };
    default:
      return state;
  }
};

export default fetchReducer;
