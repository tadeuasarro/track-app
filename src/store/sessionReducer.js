import { SET_CURRENT_USER, UPDATE_USER_TARGET } from '../actions/session';

const initialState = {
  username: false,
  id: false,
  target: false,
  token: false,
};

const sessionReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_CURRENT_USER:
      return {
        username: action.payload.user.username,
        id: action.payload.user.id,
        target: action.payload.user.target,
        token: action.payload.token,
      };
    case UPDATE_USER_TARGET:
      return {
        ...state,
        target: action.payload,
      };
    default:
      return state;
  }
};

export default sessionReducer;
