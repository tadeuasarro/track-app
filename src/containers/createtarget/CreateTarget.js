import { useSelector, useDispatch } from 'react-redux';
import { useState } from 'react';
import updateUser from '../../api/updateUser';
import { updateUserTarget } from '../../actions/session';
import { setExpenditures } from '../../actions/expenditure';
import Error from '../../components/error/Error';
import './createtarget.css';
import { setLoading, setErrors } from '../../actions/fetch';

const CreateTarget = () => {
  const { session, fetch } = useSelector(state => state);
  const { target, id } = session;
  const dispatch = useDispatch();

  const [newTarget, setNewTarget] = useState(0);

  const handleChange = e => {
    setNewTarget(e.target.value);
  };

  const handleClick = async key => {
    dispatch(setLoading());

    let res = {};

    if (!key) {
      res = await updateUser({ target: 0 }, id);
    } else {
      res = await updateUser({ target: newTarget }, id);
    }

    dispatch(setErrors(res.error));

    if (!res.error) {
      dispatch(updateUserTarget(res.payload.target));
      dispatch(setExpenditures(res.payload.expenditures));
    }
  };

  if (fetch.loading) {
    return (
      <div className="loading-container">
        <h1>Please wait...</h1>
      </div>
    );
  }

  const errorObj = (!fetch.errors ? {} : fetch.errors);

  if (target) {
    return (
      <button type="button" onClick={() => handleClick(false)} className="delete-target">Delete target</button>
    );
  }

  return (
    <div className="create-target-container">
      <input onChange={handleChange} id="target-input" className="target-input" placeholder="Input target here" />
      <Error error={errorObj.target} />
      <button onClick={() => handleClick(true)} type="button" className="create-target">Set a new target</button>
    </div>
  );
};

export default CreateTarget;
