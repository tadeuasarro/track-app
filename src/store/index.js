import { combineReducers, createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import sessionReducer from './sessionReducer';
import expenditureReducer from './expenditureReducer';
import fetchReducer from './fetchReducer';

const store = createStore(
  combineReducers({
    session: sessionReducer,
    expenditure: expenditureReducer,
    fetch: fetchReducer,
  }),
  applyMiddleware(thunk),
);

export default store;
