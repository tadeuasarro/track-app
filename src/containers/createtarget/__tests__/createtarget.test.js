import { render, screen, act } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from '../../../store/index';
import CreateTarget from '../CreateTarget';

describe('Rendering component', () => {
  it('creates an App component', () => {
    act(() => {
      render(
        <Provider store={store}>
          <CreateTarget />
        </Provider>,
      );
    });
    // eslint-disable-next-line
    const element = screen.getByText('Set a new target');
  });
});

describe('Display', () => {
  it('renders correctly', () => {
    const comp = renderer.create(
      <Provider store={store}>
        <CreateTarget />
      </Provider>,
    ).toJSON();
    expect(comp).toMatchSnapshot();
  });
});
