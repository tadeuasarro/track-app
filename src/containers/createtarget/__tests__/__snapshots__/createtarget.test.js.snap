// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display renders correctly 1`] = `
<div
  className="create-target-container"
>
  <input
    className="target-input"
    id="target-input"
    onChange={[Function]}
    placeholder="Input target here"
  />
  
  <button
    className="create-target"
    onClick={[Function]}
    type="button"
  >
    Set a new target
  </button>
</div>
`;
